lib/tests/streamlit/elements/multiselect_test.py

```diff
@@ -269,3 +270,54 @@ def test_label_visibility_wrong_value(self):
                 "Unsupported label_visibility option 'wrong_value'. Valid values are "
                 "'visible', 'hidden' or 'collapsed'.",
             )
+
+    def test_max_selections(self):
+        st.multiselect("the label", ("m", "f"), max_selections=2)
+
+        c = self.get_delta_from_queue().new_element.multiselect
+        self.assertEqual(c.max_selections, 2)
+
+    def test_over_max_selections_initialization(self):
+        with self.assertRaises(StreamlitAPIException) as e:
+            st.multiselect(
+                "the label", ["a", "b", "c", "d"], ["a", "b", "c"], max_selections=2
+            )
+        self.assertEqual(
+            str(e.exception),
+            """
+Multiselect has 3 options selected but `max_selections`
+is set to 2. This happened because you either gave too many options to `default`
+or you manipulated the widget's state through `st.session_state`. Note that
+the latter can happen before the line indicated in the traceback.
+Please select at most 2 options.
+""",
+        )
+
+    def test_max_selections_initialization_greater_than_0(self):
+        with self.assertRaises(StreamlitAPIException) as e:
+            st.multiselect(
+                "the label", ["a", "b", "c", "d"], ["a", "b", "c"], max_selections=0
+            )
+        self.assertEqual(
+            str(e.exception),
+            """
+Multiselect has 3 options selected but `max_selections`
+is set to 0. This happened because you either gave too many options to `default`
+or you manipulated the widget's state through `st.session_state`. Note that
+the latter can happen before the line indicated in the traceback.
+Please select at most 0 options.
+""",
+        )
+
+    @parameterized.expand(
+        [
+            (["a", "b", "c"], 3),
+            (["a"], 1),
+            ([], 0),
+            ("a", 1),
```

Comment:
So here's some really funky behavior that I just realized is currently a thing that we might want to think about a bit on the product side (CC @jrieke). I think the answer is that we likely can't change anything about this right now because the change will be breaking, but at some point we'll want to revisit making this more sane. Thankfully, this is enough of an edge case that I doubt people actually run into this in practice when writing sane code.

Guess what the following multiselect is initialized with on the frontend:
```python
st.multiselect("oh nooooo", ["a", "b", "c", "wow"], default="abc")
```

The answer is that Python's `str` type is both `Iterable` and a `Sequence` (strings are just collections of characters, after all), ... so you get a multiselect with "a", "b", and "c" selected.

On the other hand,
```python
st.multiselect("oh nooooo", ["w", "o", "w", "wow"], default="wow")
```
in this case we preselect "wow" even though strings are both `Iterable` and `Sequences` because we find `"wow"` in the options list 😱 

Not sure if there's anything we can really do here at this point, but wanted to point out this behavior.

Reply:
Hah! Good catch @vdonato. I mean we could explicitly check if `default` is a single string and if it is, only treat is as such, and not as an Iterable. I'm going back and forth between "this is only bug so it's ok if we break something" and "this *could* also be used as a feature". Anyway, I suspect nobody uses this in practice, so probably not worth spending too many brain cycles on it right now. We could also just fix it in 2.0... 

Can you create a Github issue and link it in [the 2.0 doc](https://www.notion.so/streamlit/Streamlit-2-0-idea-dump-7ab8221250ba459a9edbebab5e1551ec) so we keep track of it?

Reply:
Will do. Yeah, I wanted to point this strangeness out after finding it but agree that it's such an obscure edge case that I doubt we want to put many cycles into figuring out what to do with it right now